### PR TITLE
Add opt-in option to auto run "make checkstatic" locally upon git push

### PR DESCRIPTION
**Pull request by johnsaigle**
## Brief summary of changes

This PR adds a git hook that will run `make checkstatic` for you when you do `git push`. It will stop the push from going through if `make checkstatic` fails. The idea is to encourage developers to fix static analysis errors locally before pushing to GitHub. This should help accelerate code review and prevent clogging up Travis. 

This is opt-in; you can enable it by running `make githooks` just once.

## Testing

1. Checkout this branch and run `make githooks`
2. Create a new branch and try to do a `git push`. `make checkstatic` should trigger.

## Related

This is a remake of #5246 that using plain git hooks instead of requiring a 3rd-party library.

### CODE REVIEW

**Comment by johnsaigle** [comment]
@maltheism Could you add Passed Manual Tests if this worked for you?

File: Makefile [hunk 1]
```diff
@@ -32,3 +32,8 @@ unittests: phpdev
 
 # Perform all tests that don't require an install.
 check: checkstatic unittests
+
+# Configure githooks for this repository to be .githooks instead of the default
+# .git/hooks/ (which is ignored).
```
Comment (driusan) [inline Makefile]: This belongs in the documentation, not the Makefile

Comment (johnsaigle) [reply Makefile]: I changed it to "Configures..." to make it more clear that the comment is describing what the command does, not telling the user to do an additional action.

**Comment by maltheism** [comment]
Hi @johnsaigle, I just tried to manual test the PR. I checked out your PR and make a new branch based on it. Then I made a small change to a file and committed it. Lastly I attempted the git push -u origin test_githooks and here is my output:
```
mademoiselle:Loris alizee$ git push -u origin test_githooks
Enumerating objects: 29, done.
Counting objects: 100% (29/29), done.
Delta compression using up to 8 threads
Compressing objects: 100% (21/21), done.
Writing objects: 100% (24/24), 33.04 KiB | 3.67 MiB/s, done.
Total 24 (delta 13), reused 0 (delta 0)
remote: Resolving deltas: 100% (13/13), completed with 5 local objects.
remote: This repository moved. Please use the new location:
remote:   https://github.com/maltheism/Loris.git
remote: 
remote: Create a pull request for 'test_githooks' on GitHub by visiting:
remote:      https://github.com/maltheism/Loris/pull/new/test_githooks
remote: 
To https://github.com/intralizee/Loris.git
 * [new branch]          test_githooks -> test_githooks
Branch 'test_githooks' set up to track remote branch 'test_githooks' from
```
The make checkstatic didn't seem to execute.

**Comment by johnsaigle** [comment]
I think that's an error on your side:

>remote: This repository moved. Please use the new location:
>remote:   https://github.com/maltheism/Loris.git
...
>To https://github.com/intralizee/Loris.git

I think it has to do with your repository being changed from `intralizee` to `maltheism`

If you Google "This repository moved. Please use the new location:" there seems to be an easy fix.

**Comment by maltheism** [comment]
Hey @johnsaigle, I updated the url and I'm now getting this output:
```
mademoiselle:Loris alizee$ git push -u origin test_git_hooks
Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Delta compression using up to 8 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 303 bytes | 303.00 KiB/s, done.
Total 3 (delta 2), reused 0 (delta 0)
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
remote: 
remote: Create a pull request for 'test_git_hooks' on GitHub by visiting:
remote:      https://github.com/maltheism/Loris/pull/new/test_git_hooks
remote: 
To https://github.com/maltheism/Loris.git
 * [new branch]          test_git_hooks -> test_git_hooks
Branch 'test_git_hooks' set up to track remote branch 'test_git_hooks' from 'origin'.

```
Although I'm still thinking it might be linked to the earlier problem. I'm going to debug my end some more.

**Comment by johnsaigle** [comment]
That looks like a successful push to me.

Did you run `make githooks` in the testing instructions after checking out this branch?

**Comment by maltheism** [comment]
@johnsaigle That was it! Works and you helped solve an issue I didn't even know about until now. 😆

**Comment by johnsaigle** [comment]
Restart travis after release testing

File: Makefile (hunk 1 above)
Comment (driusan) [reply Makefile]: But the user should be told to do an additional action in the documentation if they so desire.. we shouldn't be changing their git configuration in the Makefile

**Comment by driusan** [comment]
This is also going to the wrong branch

File: Makefile (hunk 1 above)
Comment (ridz1208) [reply Makefile]: @johnsaigle documentation for this could be added in the developer section of the new wiki

Comment (johnsaigle) [reply Makefile]: I added a note to CONTRIBUTING since we don't have a document for this kind of thing on the new Wiki yet

**Comment by ridz1208** [comment]
@johnsaigle I'm getting the same issue as @maltheism 
```
rabouh@ridz-UbuntuMac:/var/www/loris$ git push origin bla
Counting objects: 14, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (7/7), done.
Writing objects: 100% (14/14), 1.41 KiB | 1.41 MiB/s, done.
Total 14 (delta 7), reused 9 (delta 4)
remote: Resolving deltas: 100% (7/7), completed with 3 local objects.
remote: 
remote: Create a pull request for 'bla' on GitHub by visiting:
remote:      https://github.com/ridz1208/Loris/pull/new/bla
remote: 
To github.com:ridz1208/Loris.git
 * [new branch]          bla -> bla
```

**Comment by johnsaigle** [comment]
Did you do `make githooks`?

File: Makefile (hunk 1 above)
Comment (driusan) [reply Makefile]: The CONTRIBUTING shouldn't say "run make githooks". "githooks" isn't a reasonable make target. The CONTRIBUTING.md should say "run git config core.hooksPath .githooks or copy .githooks/pre-push to your .git/hooks directory"

Comment (ridz1208) [reply Makefile]: @driusan why ? because it's a one time configuration ?

what makes a reasonable make target ?

Comment (driusan) [reply Makefile]: because it's configuring the user's software (and doing it in an opinionated way), not building the software. Make is a build tool.

**Comment by ridz1208** [comment]
This works, only 1 change requested by @driusan needs to be addressed. see comment from October 9th

https://github.com/aces/Loris/pull/5263#pullrequestreview-299435324

(github is not displaying the changes requested flag)

**Pull request merged**